// @takram/three-atmosphere/build/index.js
import { A as U, c as P, d as X, j as g, I as h, M as y, P as x, p as H, o as L, k as v, l as B, m as Z, S as k, n as w, r as b, u as d, w as z, x as F, z as V, C as O, q as W, T as Y, a as j, b as q, g as K, i as J, f as Q, h as $, e as ee, s as te, t as ae, v as re, y as se, B as oe } from "./shared.js";
import { Color as R, Vector3 as S, Matrix3 as m } from "three";
import { saturate as i, clamp as u } from "@takram/three-geospatial";
const l = /* @__PURE__ */ new S(), C = /* @__PURE__ */ new m(
  3.2404542,
  -1.5371385,
  -0.4985314,
  -0.969266,
  1.8760108,
  0.041556,
  0.0556434,
  -0.2040259,
  1.0572252
);
function D(t, s = new R()) {
  const a = t, o = a ** 2, T = (0.860117757 + 154118254e-12 * a + 128641212e-15 * o) / (1 + 842420235e-12 * a + 708145163e-15 * o), E = (0.317398726 + 422806245e-13 * a + 420481691e-16 * o) / (1 - 289741816e-13 * a + 161456053e-15 * o), c = 3 * T / (2 * T - 8 * E + 4), r = 2 * E / (2 * T - 8 * E + 4), n = 1, _ = r > 0 ? c * n / r : 0, I = r > 0 ? (1 - c - r) * n / r : 0, e = l.set(_, n, I).applyMatrix3(C);
  return e.x = i(e.x), e.y = i(e.y), e.z = i(e.z), s.setFromVector3(e.normalize());
}
function A(t) {
  return 4600 * (1 / (0.92 * u(t, -0.4, 2) + 1.7) + 1 / (0.92 * t + 0.62));
}
function M(t, s = new R()) {
  return D(
    A(t),
    s
  );
}
export {
  U as AerialPerspectiveEffect,
  P as AtmosphereMaterialBase,
  X as AtmosphereParameters,
  g as IRRADIANCE_TEXTURE_HEIGHT,
  h as IRRADIANCE_TEXTURE_WIDTH,
  y as METER_TO_UNIT_LENGTH,
  x as PrecomputedTexturesLoader,
  H as SCATTERING_TEXTURE_DEPTH,
  L as SCATTERING_TEXTURE_HEIGHT,
  v as SCATTERING_TEXTURE_MU_SIZE,
  B as SCATTERING_TEXTURE_MU_S_SIZE,
  Z as SCATTERING_TEXTURE_NU_SIZE,
  k as SCATTERING_TEXTURE_R_SIZE,
  w as SCATTERING_TEXTURE_WIDTH,
  b as SKY_RENDER_ORDER,
  d as SkyLightProbe,
  z as SkyMaterial,
  F as StarsGeometry,
  V as StarsMaterial,
  O as SunDirectionalLight,
  W as TRANSMITTANCE_TEXTURE_HEIGHT,
  Y as TRANSMITTANCE_TEXTURE_WIDTH,
  j as aerialPerspectiveEffectOptionsDefaults,
  q as atmosphereMaterialParametersBaseDefaults,
  M as convertBVIndexToLinearSRGBChromaticity,
  D as convertTemperatureToLinearSRGBChromaticity,
  K as getECIToECEFRotationMatrix,
  J as getMoonDirectionECEF,
  Q as getMoonDirectionECI,
  $ as getSunDirectionECEF,
  ee as getSunDirectionECI,
  te as getSunLightColor,
  ae as skyLightProbeParametersDefaults,
  re as skyMaterialParametersDefaults,
  se as starsMaterialParametersDefaults,
  oe as sunDirectionalLightParametersDefaults
};
